FeedStoreChallenge/CoreDataFeedStore/CoreDataFeedStore.swift

```diff
@@ -0,0 +1,110 @@
+//
+//  CoreDataFeedStore.swift
+//  FeedStoreChallenge
+//
+//  Created by Danil Lahtin on 25.04.2021.
+//  Copyright © 2021 Essential Developer. All rights reserved.
+//
+
+import Foundation
+import CoreData
+
+public final class CoreDataFeedStore: FeedStore {
+    private let context: NSManagedObjectContext
+
+    public init(storeUrl: URL, bundle: Bundle = .main) throws {
+        let container = try NSPersistentContainer.load(modelName: "ManagedCacheModel", at: storeUrl, in: bundle)
+        context = container.newBackgroundContext()
+    }
+
+    public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
+        context.modify(ManagedCache.removeIfExists, completion: completion)
+    }
+
+    public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
+        context.modify({ context in
+            try ManagedCache.removeIfExists(in: context)
+            let cache = ManagedCache(with: feed, timestamp: timestamp, context: context)
+            context.insert(cache)
+        }, completion: completion)
+    }
+
+    public func retrieve(completion: @escaping RetrievalCompletion) {
+        context.perform { [context] in
+            do {
+                let cache = try ManagedCache.request(in: context)
+                let result: RetrieveCachedFeedResult = cache.map({
+                    .found(feed: $0.makeLocalFeedImages(), timestamp: $0.timestamp)
+                }) ?? .empty
+
+                completion(result)
+            } catch {
+                completion(.failure(error))
+            }
+        }
+    }
+}
+
+// MARK: - Helpers
+
+private extension NSManagedObjectContext {
+    func modify(_ block: @escaping (NSManagedObjectContext) throws -> (), completion: @escaping (Error?) -> ()) {
+        perform { [self] in
+            do {
+                try block(self)
+                try self.save()
```

Comment:
Some operations may not change the context at all (e.g., trying to delete a cache that is already empty). Calling save wouldn't be needed in those cases, but this method doesn't have knowledge about the operations to decide when to call save or not. So you can check if there are changes in the context before calling save:

```suggestion
                try block(self)
                if self.hasChanges {
                        try self.save()
                }
```

From the docs:

*Always verify that the context has uncommitted changes (using the hasChanges property) before invoking the save: method. Otherwise, Core Data may perform unnecessary work.*

https://developer.apple.com/documentation/coredata/nsmanagedobjectcontext/1506866-save